Use .toJSON() (& delete ._rev) rather than .changed

timblack1 opened the pull request:
This works better for me than the change in #11, because for some reason in my case modelOrCollection.changed doesn't contain all (or sometimes any) of the changed attributes.

timblack1 committed:
Use .toJSON() (& delete ._rev) rather than .changed

This works better for me than the change in #11, because for some reason in my case modelOrCollection.changed doesn't contain all (or sometimes any) of the changed attributes.

gr2m commented:
Thanks for the pull request, Tim! It's a vague though, I wouldn't like to merge a potential change of behaviour based on assumptions. Can you investigate when `modelOrCollection.changed` does not have the changes, and why you remove the `_rev` property?

timblack1 commented:
Here is some code which demonstrates the kind of problem I'm having:

``` JavaScript
var m = new Backbone.Model()
m.on('change', function(data){ console.log(m.changed) })
m.set('test','testing')
// Console output:  Object {test: "testing"}
m.set('test2','testing2')
// Console output:  Object {test2: "testing2"}
m.save()
// backbone-hoodie only saves {test2: "testing2"}, and fails to save {test: "testing"}.
// This is unexpected behavior!
```

So I think backbone-hoodie needs to save m.toJSON() rather than m.changed, in order not to save all attributes which have changed since the last call to m.set(), but all attributes which have changed since the last call to m.save().  An alternative would be to make backbone-hoodie automatically do something like the following to call m.save() after every m.set(), but I don't think you want this:

``` JavaScript
m.on('change', function(){
  m.save()
}
```

What is more, this line from Backbone's docs troubles me, since I'm not calling m.save() within a change event:  "Note that this method, and the following change-related ones, are only useful during the course of a 'change' event."  Is m.changed guaranteed to contain the data you need after the change event has finished occurring?

I decided to remove the _rev property because of this comment:  https://github.com/hoodiehq/backbone-hoodie/issues/11#issue-31194038.  But I didn't test whether leaving the _rev property in attributes causes a problem.

gr2m commented:
Thanks Tim, that is really helpful. I didn't run into this myself as I save every change (because why not :) but I agree that it's a bug that changes do not get persisted when people use `.set`.

I'll think about it a bit more next week, there was a reason why I used the `.changed` property, just can't recall it right now

gr2m commented:
@timblack1 Okay, it looks good to me. I don't have any tests right now to give this a spin and to assure that it doesn't break any other stuff. Do you have it in use in an app yet?

timblack1 commented:
@gr2m I'm using it in https://github.com/timblack1/heaven-and-hell-machine, but haven't deployed that app to production yet.  I don't have tests for it yet either, sorry.

gr2m commented:
@timblack1 just fyi I'll put this into https://github.com/gr2m/hoodie-plugin-backbone, which is going to be the successor of this repository

timblack1 commented:
Ok, thanks @gr2m.  How can I tell whether https://github.com/gr2m/hoodie-plugin-backbone is ready for production?  In other words, is it the successor of this repository yet, or if not, when will it be?

gr2m commented:
@timblack1 it is ready, I still need to update this repository's README, it's on my list :)
